google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerPool.java

```diff
@@ -337,6 +337,9 @@ private void initialize() {
 
   @VisibleForTesting
   Spanner createSpanner(SpannerPoolKey key, ConnectionOptions options) {
+    if (options.isEnabledOpenTelemetryMetrics() && !ConnectionSpannerOptions.isEnabledOpenTelemetryMetrics()) {
+      ConnectionSpannerOptions.enableOpenTelemetryMetrics();
```

Comment:
This is a static method that enables the option globally. Theoretically, we could have the following situation:

1. Someone creates a `Connection` with `enableOpenTelemetryMetrics=false`. That means that a `Spanner` instance without an `OpenTelemetry` will be created.
2. Someone then creates another connection with `enableOpenTelemetryMetrics=true`. That in the current implementation will cause `ConnectionSpannerOptions.enableOpenTelemetryMetrics()` to be enabled, but as that property is not part of the `SpannerPoolKey`, no new `Spanner` instance will be created. Meaning that the `Connection` will not collect any metrics.

However, I also expect problems if you do add the property `enableOpenTelemetryMetrics` to the `SpannerPoolKey`, because then this method will again be called, even though there are already a `Spanner` instance that has been created without an `OpenTelemetry` instance.

So what we need to do is:
1. Add `enableOpenTelemetryMetrics` as an attribute to the `SpannerPoolKey`
2. Add tests that verify that all combinations of creating connections with and without metrics actually create the type of connection that they should, and that it works with the global `ConnectionSpannerOptions.enableOpenTelemetryMetrics()` setting.

Reply:
I think the main problem is once we enable `SpannerOptions.enableOpenTelemetryMetrics` to true globally, we do not have a way to disable it. 

What if someone creates a `Connection` with `enableOpenTelemetryMetrics=true`. And then, someone creates a Connection with `enableOpenTelemetryMetrics=false`, who will get a `Spanner` instance with` OpenTelemetry` enabled. 

Multiple `Spanner` instances would share the same global `ConnectionSpannerOptions.enableOpenTelemetryMetrics()` setting.

Reply:
Should we make `enableOpenTelemetryMetrics` in `ConnectionOptions` also be a global setting? If the caller of Connection API wants to enable OpenTelemetry metrics instead of putting this as a parameter in URI, they must call: 

```java
ConnectionOptions.enableOpenTelemetryMetrics();
...
ConnectionOptions.Builder connectionOptionsBuilder = ConnectionOptions.newBuilder().setUri(uri);

connectionOptionsBuilder.setOpenTelemetry(openTelemetry);

ConnectionOptions connectionOptions = connectionOptionsBuilder.build();
Connection spannerConnection = connectionOptions.getConnection();
```

Reply:
TLDR:

We should:
1. Add a `ConnectionOptions$Builder#setOpenTelemetry(OpenTelemetry openTelemetry)` method
2. Add this `ConnectionOptions.openTelemetry` property to the `SpannerPoolKey`, and set the `OpenTelemetry` instance on the `Spanner` instance that is built.
3. Not add a global `ConnectionOptions.enableOpenTelemetryMetrics()`. We don't need it. The users of the Connection API (JDBC + PGAdapter) should just call `SpannerOptions.enableOpenTelemetryMetrics()`.

Longer version:

First to clarify one thing: The `SpannerOptions.enableOpenTelemetryMetrics()` method only instructs the Java client to use OpenTelemetry __instead of__ OpenCensus for metrics. It does not itself create and register an OpenTelemetry instance etc.

We don't need to replicate that in the Connection API. Instead, a user of the Connection API should itself just call `SpannerOptions.enableOpenTelemetryMetrics()` directly before trying to use OpenTelemetry. This is probably something that they should do in any case, as none of them want to use OpenCensus. Note here that 'a user of the Connection API' in reality means 'JDBC' or 'PGAdapter'.

Regarding how to set the `OpenTelemetry` instance for a connection:
1. Adding a `ConnectionOptions$Builder#setOpenTelemetry(OpenTelemetry openTelemetry)` method makes sense. That property should then be propagated to the `SpannerOptions$Builder#setOpenTelemetry(...)` method.
4. That then also means that this new `OpenTelemetry` property needs to be added to the `SpannerPoolKey`, so that if two connections with different `OpenTelemetry` instances are created, they also get two different `Spanner` instances.

-----

The bigger challenge is how to create and configure the `OpenTelemetry` instance that should be set on a `ConnectionOptions` instance by the JDBC driver and PGAdapter. Note here that users of JDBC and PGAdapter do not always have programmatic access to the configuration (and none of them have access to `ConnectionOptions`). That is not something that needs to be solved in this PR, but we should make sure that we don't make a choice here that makes it impossible to use in JDBC and PGAdapter.

For PGAdapter, we have solved this by using the [AutoConfiguredOpenTelemetrySdk](https://github.com/GoogleCloudPlatform/pgadapter/blob/7af6e01aa1683381e254ab6860103898a636107d/src/main/java/com/google/cloud/spanner/pgadapter/Server.java#L107). This library reads configuration properties from System properties and/or environment variables that should be used for OpenTelemetry, and configures OpenTelemetry based on that. This OpenTelemetry instance can then be used for the `ConnectionOptions` instance that is created by PGAdapter.

For JDBC, I could imagine a couple of different options (all in parallel possible):
1. A public static method that sets a global `OpenTelemetry` object that should be used for all JDBC connections in the current JVM. So something like `JdbcDriver.setOpenTelemetry(OpenTelemetry openTelemetry)`. This method has to be called before creating any connections.
2. The standard way to get a JDBC connection is to call `DriverManager.getConnection(String url[, Properties properties])`. The `properties` map can also contain complex objects like `OpenTelemetry`. A user could then supply a specific `OpenTelemetry` that should be used for a connection in this `properties` argument. This would only be available to users who are using the JDBC driver directly, and not for users who are using for example Hibernate.
5. An `autoConfigOpenTelemetry=true` connection URL property. Setting this to true will activate similar logic as described for PGAdapter above: Using the `AutoConfigureOpenTelemetrySdk` an `OpenTelemetry` instance will be created and configured based on system properties and environment variables.

In all the above cases, the `OpenTelemetry` instance that has been configured will be passed on to the `ConnectionOptions` instance that is built.

Reply:
Fixed. I wonder if JDBC should use the same approach as PGAdapter.